File: solutions/elixir/resistor-color-trio/2/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @resistors %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @scale [:ohms, :kiloohms, :megaohms, :gigaohms]

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors), do: colors |> Enum.take(3) |> do_label()

  defp do_label(colors, acc \\ 0)
  defp do_label([:black], 0), do: {0, :ohms}

  defp do_label([color], acc) do
    ohms = acc * 10 ** @resistors[color]

    Enum.reduce_while(@scale, ohms, fn unit, acc ->
      if div(acc, 1000) * 1000 == acc do
        {:cont, div(acc, 1000)}
      else
        {:halt, {acc, unit}}
      end
    end)
  end

  defp do_label([color | rest], acc), do: do_label(rest, acc * 10 + @resistors[color])
end

File: solutions/lua/resistor-color-trio/2/resistor-color-trio.lua
local resistors = {
	black = 0,
	brown = 1,
	red = 2,
	orange = 3,
	yellow = 4,
	green = 5,
	blue = 6,
	violet = 7,
	grey = 8,
	white = 9,
}

local scale = { "ohms", "kiloohms", "megaohms", "gigaohms" }

return {
	decode = function(c1, c2, c3)
		local value = (10 * resistors[c1] + resistors[c2]) * 10 ^ resistors[c3]

		for _, unit in ipairs(scale) do
			if value % 1000 == value then
				return value, unit
			else
				value = value / 1000
			end
		end
	end,
}
